File: Leetcode/Javascript/39. Combination Sum.js
var combinationSum = function(candidates, target) {
    candidates.sort(function(a,b) {
        return a-b;
    });
    var result = [];
    dfs(0, target, candidates, [], result);
    return result;
};
function dfs(index, remain, nums, list, result) {
    if(remain < 0) return;
    if(!remain) result.push(list.slice());
    for(var i = index; i < nums.length; ++i) {
        list.push(nums[i]);
        dfs(i, remain-nums[i], nums, list, result);
        list.pop();
    }
}

File: Leetcode/Three/39. Combination Sum.java
public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        dfs(result, 0, candidates, target, new ArrayList<Integer>());
        return result;
    }
    private void dfs(List<List<Integer>> result, int index, int[] candidates, int target, List<Integer> line) {
        if(target < 0) return;
        else if(target == 0) {
            result.add(new ArrayList<Integer>(line));
            return;
        }
        else {
            for(int i = index; i < candidates.length && candidates[i] <= target; ++i) {
                line.add(candidates[i]);
                dfs(result, i, candidates, target-candidates[i], line);
                line.remove(line.size()-1);
            }
        }
    }
}
